overseer/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+/// messages in the system.
+///
+/// [`AllMessages`]: enum.AllMessages.html
+/// [`Subsystem`]: trait.Subsystem.html
+/// [`Overseer`]: struct.Overseer.html
+enum ToOverseer {
+	/// This is a message sent by a `Subsystem`.
+	SubsystemMessage(AllMessages),
+
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	SpawnJob {
+		s: BoxFuture<'static, ()>,
+		res: oneshot::Sender<SubsystemResult<()>>,
+	},
+}
+
+/// Some event from outer world.
+enum Event {
+	BlockImport,
+	BlockFinalized,
+	MsgToSubsystem(AllMessages),
+	Stop,
+}
+
+/// Some message that is sent from one of the `Subsystem`s to the outside world.
+pub enum OutboundMessage {
+	SubsystemMessage {
+		msg: AllMessages,
+	}
+}
+
+/// A handler used to communicate with the [`Overseer`].
+///
+/// [`Overseer`]: struct.Overseer.html
+pub struct OverseerHandler {
+	events_tx: mpsc::Sender<Event>,
+}
+
+impl OverseerHandler {
+	/// Inform the `Overseer` that that some block was imported.
+	pub async fn block_imported(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::BlockImport).await?;
+
+		Ok(())
+	}
+
+	/// Send some message to one of the `Subsystem`s.
+	pub async fn send_msg(&mut self, msg: AllMessages) -> SubsystemResult<()> {
+		self.events_tx.send(Event::MsgToSubsystem(msg)).await?;
+
+		Ok(())
+	}
+
+	/// Inform the `Overseer` that that some block was finalized.
+	pub async fn block_finalized(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::BlockFinalized).await?;
+
+		Ok(())
+	}
+
+	/// Tell `Overseer` to shutdown.
+	pub async fn stop(&mut self) -> SubsystemResult<()> {
+		self.events_tx.send(Event::Stop).await?;
+
+		Ok(())
+	}
+}
+
+impl Debug for ToOverseer {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			ToOverseer::SubsystemMessage(msg) => {
+				write!(f, "OverseerMessage::SubsystemMessage({:?})", msg)
+			}
+			ToOverseer::SpawnJob { .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+	}
+}
+
+/// A running instance of some [`Subsystem`].
+///
+/// [`Subsystem`]: trait.Subsystem.html
+struct SubsystemInstance<M: Debug> {
+	tx: mpsc::Sender<FromOverseer<M>>,
+}
+
+/// A context type that is given to the [`Subsystem`] upon spawning.
+/// It can be used by [`Subsystem`] to communicate with other [`Subsystem`]s
+/// or to spawn it's [`SubsystemJob`]s.
+///
+/// [`Overseer`]: struct.Overseer.html
+/// [`Subsystem`]: trait.Subsystem.html
+/// [`SubsystemJob`]: trait.SubsystemJob.html
+pub struct SubsystemContext<M: Debug>{
+	rx: mpsc::Receiver<FromOverseer<M>>,
+	tx: mpsc::Sender<ToOverseer>,
+}
+
+/// A signal used by [`Overseer`] to communicate with the [`Subsystem`]s.
+///
+/// [`Overseer`]: struct.Overseer.html
+/// [`Subsystem`]: trait.Subsystem.html
+#[derive(Debug)]
+pub enum OverseerSignal {
+	/// `Subsystem` should start working.
+	StartWork,
```

Comment:
From the guide: 

```rust
enum OverseerSignal {
  /// Signal to start work localized to the relay-parent hash.
  StartWork(Hash),
  /// Signal to stop (or phase down) work localized to the relay-parent hash.
  StopWork(Hash),
}
```

I see these signals are not yet sent in the code, that's fine for a follow-up.